### PR TITLE
Refactor Review Moderation logic to decouple ReviewListFragment and ReviewDetailFragment

### PR DESCRIPTION
**Pull request by hichamboushaba**
Part of: #5647

⚠️ Please don't merge, the FluxC PR needs to be merged first: https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/2316

Only one reviewer is needed, but feel free to review if you have time 🙏  

### Description
The review moderation currently is causing a tight couple between `ReviewListFragment` and `ReviewDetailFragment`, which prevents reusing the same logic in `ProductReviewsFragment`. The code also is hard to follow, it uses EventBus notifications, and the logic is spread across ViewModel and UI (Fragment and Adapter).
A temporary solution was to hide the moderation buttons when viewing the opening `ReviewDetailFragment` from `ProductReviewsFragment` (check https://github.com/woocommerce/woocommerce-android/pull/3003).

This is a first PR in the refactoring, it brings just the same functionality as before, which means successive moderation requests are not handled well (before the moderation request was lost if a new one was submitted before the first one is submitted, while now there is just a UI issue where the shown status will be incorrect during a short time).
The functionality will be improved in a second PR, and a third PR will bring moderation to `ProductReviewsFragment`.

### Architecture
The bulk of the logic is part of the `ReviewModerationHandler` class which has the following API:

```kotlin
class ReviewModerationHandler {
   val pendingModerationStatus: Flow<ReviewModerationStatus>
   fun postModerationRequest(review: ProductReview, newStatus: ProductReviewStatus)
   fun undoLastOperation()
}
```
And to help reuse code for ViewModels and Fragments, and enforce a contract, I added two interfaces:
1. `ReviewModerationConsumer` to be implemented by ViewModels who want to handle moderation events.
2. `ReviewModerationUi` to be implemented by corresponding Fragments.
If you can think of any improvement to the architecture, please let me know 🙏

### Testing instructions

1. Create some reviews on your store.
3. Open the Reviews list screen on the app.
4. Tap on one of the reviews.
5. Click on one of the moderation options (SPAM and TRASH should trigger a removal of the review from the list).
6. Wait until the snackbar is hidden.
7. Confirm the review status has been correctly updated.
8. Repeat step 4.
9. Click on "Undo" in the snackbar.
10. Confirm the action is canceled.

- [x] I have considered if this change warrants user-facing release notes and have added them to RELEASE-NOTES.txt if necessary.

### CODE REVIEW

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/reviews/ReviewListFragment.kt [hunk 1]
```diff
@@ -346,14 +251,14 @@ class ReviewListFragment :
                     review.remoteId,
                     launchedFromNotification = false,
                     enableModeration = true,
-                    tempStatus = pendingModerationNewStatus
+                    tempStatus = review.status
```
Comment (hichamboushaba) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/reviews/ReviewListFragment.kt:254]: Since the reviews we expose from the ViewModel have the correct statuses, we can just read it from the object itself.
I may update this in the next PR to read it from the `ReviewModerationHandler` directly instead of passing an addition navArg 🤔

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/reviews/ReviewListViewModel.kt [hunk 2]
```diff
@@ -27,48 +23,46 @@ import com.woocommerce.android.viewmodel.LiveDataDelegate
...
+    val reviewList = _reviewList.combineWithModerationStatus(pendingReviewModerationStatus)
 
-    final val viewStateData = LiveDataDelegate(savedState, ViewState())
+    val viewStateData = LiveDataDelegate(savedState, ViewState())
     private var viewState by viewStateData
 
     init {
         EventBus.getDefault().register(this)
         dispatcher.register(this)
+        observeReviewUpdates()
```
Comment (hichamboushaba) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/reviews/ReviewListViewModel.kt:64]: Moving this here is not relevant to this PR, but I noticed it was called in `start` which gets called on each `onViewCreated`, which means we were having multiple Flow collections for the same goal if we navigate to another screen and come back and on screen rotations.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/reviews/ReviewListViewModel.kt [hunk 3]
```diff
@@ -217,6 +181,7 @@ class ReviewListViewModel @Inject constructor(
     private fun observeReviewUpdates() {
         viewModelScope.launch {
             unseenReviewsCountHandler.observeUnseenCount()
+                .distinctUntilChanged()
```
Comment (hichamboushaba) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/reviews/ReviewListViewModel.kt]: I added this to prevent a useless review fetching when we update the review status.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/reviews/ReviewModerationHandler.kt [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+import javax.inject.Singleton
+
+@Singleton
+class ReviewModerationHandler @Inject constructor(
+    private val selectedSite: SelectedSite,
+    private val productStore: WCProductStore,
+    @AppCoroutineScope private val appCoroutineScope: CoroutineScope
+) {
+    companion object {
+        const val UNDO_DELAY = 2750L
```
Comment (hichamboushaba) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/reviews/ReviewModerationHandler.kt:27]: This duration matches the [material duration](https://github.com/material-components/material-components-android/blob/dce4419744ba4e336e31a80521ab8a9ecb1136de/lib/java/com/google/android/material/snackbar/SnackbarManager.java#L32) for SnackBars with `LENGTH_LONG`, so it's matching the behavior we had before, but while being testable instead of UI controlled.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/reviews/ReviewModerationConsumer.kt [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+import com.woocommerce.android.ui.reviews.ProductReviewStatus.SPAM
+import com.woocommerce.android.ui.reviews.ProductReviewStatus.TRASH
+import com.woocommerce.android.viewmodel.combineWith
+import kotlinx.coroutines.flow.collect
+import kotlinx.coroutines.flow.filter
+
+interface ReviewModerationConsumer {
+    val ReviewModerationConsumer.reviewModerationHandler: ReviewModerationHandler
+    val ReviewModerationConsumer.rawReviewList: LiveData<List<ProductReview>>
+    fun ReviewModerationConsumer.reloadReviewsFromCache()
```
Comment (hichamboushaba) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/reviews/ReviewModerationConsumer.kt:16]: The goal of putting the properties/function as extensions, is to make them visible only to the implementing class, I mean simulating a `protected` visibility, which is not supported natively for interfaces by Kotlin.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/reviews/ReviewModerationUi.kt [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+import androidx.fragment.app.Fragment
+import androidx.lifecycle.DefaultLifecycleObserver
+import androidx.lifecycle.LifecycleOwner
+import androidx.lifecycle.LiveData
+import com.google.android.material.snackbar.Snackbar
+import com.woocommerce.android.R
+import com.woocommerce.android.model.ActionStatus
+import com.woocommerce.android.ui.base.UIMessageResolver
+
+interface ReviewModerationUi
```
Comment (hichamboushaba) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/reviews/ReviewModerationUi.kt:11]: The only goal of this is to avoid cluttering the API surface (and autocomplete suggestions) of all Fragments with the extension `observeModerationStatus`, in this way, only Fragments that implement `ReviewModerationUi` will see it.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/reviews/ReviewModerationConsumer.kt (hunk 5 above)
Comment (JorgeMucientes) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/reviews/ReviewModerationConsumer.kt:16]: Nice! TIL about using extension properties to simulate `protected` visibility. 

One question regarding this implementation. Why not use Flows everywhere and only in the "last moment" use the `asLiveData`. This will enable to keep using the same API (Flow APIs) as much as possible, and reduce the amount of `asLiveData` conversion needed.  Only when exposing a property to the View/Fragment use the `asLiveData`. In this case, only `ReviewModerationConsumer.reviewList` would be LiveData and  the rest of the properties could be Flows. This would require a small refactor in `ReviewListViewModel` to convert `_reviewList` to MutableStateFlow too. 

No need to change this, just wondering if this could remove some unnecessary calls to `asLiveData` and make future migrations or operations over flow easier.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/reviews/ReviewModerationUi.kt [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+import androidx.lifecycle.LifecycleOwner
+import com.google.android.material.snackbar.Snackbar
+import com.woocommerce.android.R
+import com.woocommerce.android.model.ActionStatus
+import com.woocommerce.android.ui.base.UIMessageResolver
+
+interface ReviewModerationUi
+
+fun ReviewModerationUi.observeModerationStatus(
+    reviewModerationViewModel: ReviewModerationConsumer,
```
Comment (JorgeMucientes) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/reviews/ReviewModerationUi.kt]: NP, while this is most likely always going to be a ViewModel I'd rename the param to `reviewModerationConsumer `

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/reviews/ReviewListViewModel.kt (hunk 3 above)
Comment (JorgeMucientes) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/reviews/ReviewListViewModel.kt]: Hmm, I think this might be redundant as `unseenReviewsCount` flow from `UnseenReviewsCountHandler` already calls `distinctUntilChanged()`

**Review by JorgeMucientes — APPROVED** [review]

**Comment by JorgeMucientes** [comment]
Really neat refactor @hichamboushaba. Super nice how things are now completely decoupled and reusable. I really like how much the View code (ReviewListAdapter and Fragment) got simplified. 

I've approved the PR and left some minor suggestions. Feel free to apply them. Great job!

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/reviews/ReviewListViewModel.kt (hunk 3 above)
Comment (hichamboushaba) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/reviews/ReviewListViewModel.kt]: Yes, you are probably right, nice catch. 
Maybe the behavior I saw was because the `observeReviewUpdates` was called in `start` which gets called multiple times, I'll remove this call and retest.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/reviews/ReviewModerationConsumer.kt (hunk 5 above)
Comment (hichamboushaba) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/reviews/ReviewModerationConsumer.kt:16]: Nice suggestion and I think it makes a lot of sense, I'll add it.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/reviews/ReviewModerationUi.kt (hunk 7 above)
Comment (hichamboushaba) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/reviews/ReviewModerationUi.kt]: > I'd rename the param to `reviewModerationViewModel`

I think you meant to suggest a different name, since it's already called this 🙂

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/reviews/ReviewModerationConsumer.kt (hunk 5 above)
Comment (hichamboushaba) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/reviews/ReviewModerationConsumer.kt:16]: Actually since I already have two pending PRs which depend on this, I will add this improvement after finishing with them, just to avoid having to deal with unnecessary conflicts.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/reviews/ReviewListViewModel.kt (hunk 3 above)
Comment (hichamboushaba) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/reviews/ReviewListViewModel.kt]: I remove the `distinctUntilChanged`, but I added instead `drop(1)` in 6011018, without this, we will call the Repository's `fetchProductReviews` two times when opening the screen, since the `observeUnseenCount()` Flow will trigger a call too.
This doesn't actually cause two API calls, because the Repository has a boolean to check if one fetch is ongoing, but still, we shouldn't call it twice.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/reviews/ReviewModerationUi.kt (hunk 7 above)
Comment (JorgeMucientes) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/reviews/ReviewModerationUi.kt]: Yeah! sorry. I edited my comment afterward. I meant rename the param to what it is: `reviewModerationConsumer`

Comment (hichamboushaba) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/reviews/ReviewModerationUi.kt]: Sure, done in 8c6a7be
